docs/syntax.md

```diff
@@ -0,0 +1,255 @@
+importspec -> '(' [ { typename ',' } typename [','] ] ')'
+```
+
+If `importspec` is omitted, then all entities specified in the module are imported; otherwise only the specified entities are imported.
+
+### Statement
+
+Statements define types, classes, instance clauses, and derive clauses.
+
+```text
+statement -> typedef
+           | classdef
+           | instanceclause
+           | deriveclause
+```
+
+#### Type definitions
+Types may be either sum types, product types, record types, or opaque types.
+
+```text
+typedef -> prodtypedef | sumtypedef |  recordtypedef | opaquetypedef
+```
+
+##### Product type definition
+A product type definition defines a new product type.
+
+```text
+prodtypedef -> 'prod' typename { varname } '=' prod
+prod        -> { typeexp }
+typeexp     -> varname
+             | longtypename
+             | '(' prod ')'
+```
+
+Product type definitions instruct the code generator to generate a product type for the target language.
+
+##### Sum type definition
+A sum type definition defines a new sum type.
+
+```text
+sumtypedef     -> 'sum' typename { varname } '=' sum
+sum            -> sumconstructor { '|' sumconstructor }
+sumconstructor -> typename prod
+```
+
+Sum type definitions instruct the code generator to generate a sum type for the target language.
+
+##### Record type definition
+A record type definition defines a new record type.
+
+```text
+recordtypedef -> 'record' typename { varname } '=' record
+record        -> '{' [ field { ',' field  } ] '}'
+field         -> fieldname ':' prod
+````
+
+Record type definitions instruct the code generator to generate a record type for the target language.
+
+##### Opaque type definition
+An opaque type definition defines a new opaque type.
+
+```text
+opaquetypedef -> 'opaque' typename { varname }
+```
+
+Opaque type definitions do not instruct the code generator to generate code, and an opaque type must be instead implemented in the target language.
+
+#### Class definition
+A class definition introduces a new class.
+
+```text
+classdef       -> 'class' [ constraintexps '<=' ] classname { varname }
+constraintexp  -> classref { varname }
+                | '(' constraintexps ')'
+constraintexps -> [ constraintexp { ',' constraintexp } ]
+```
+
+Class definitions do not instruct the code generator to generate code, but
+    instead provides a means to communicate with the code generator the
+    instances one would like to generate (via a derive clause).
+
+#### Instance clause
+An instance clause specifies a type is an instance of a class.
+
+```text
+instanceclause -> 'instance'  constraint [ ':-' constraintexps ]
+constraint     -> classref { typeexp }
+```
+
+Instance clauses do not instruct the code generator to generate code, but
+    instead instructs the compiler (semantic checking) that the target language
+    provides instances for the given type provided that the given `constraintexps`
+    have instances.
```

Comment:
Excellent summary, I would perhaps say (because instances are vary abstract word):

`
that the target language environment provides type class implementations for the given type (provided that the given `constraintexps` also have implementations).
`